http-kit/src/io/pedestal/http/http_kit.clj

```diff
@@ -74,18 +77,18 @@
                              :port port})
         *server      (atom nil)
         root-handler (fn [request]
-                       ;; TODO: something like stylobate,
-                       (let [{:keys [uri]} request
-                             request' (assoc request :path-info uri)
+                       (let [{:keys [uri async-channel]} request
+                             request'       (assoc request :path-info uri)
                              *async-channel (atom nil)
                              interceptors'  (into [(async-responder *async-channel)
                                                    response-converter]
                                                   interceptors)
                              context        (-> initial-context
-                                                (assoc :request request')
+                                                (assoc :request request'
+                                                       :websocket-channel-source async-channel)
                                                 (chain/on-enter-async (fn [_]
-                                                                        (reset! *async-channel (or (:async-channel request)
-                                                                                                   (throw (ex-info "No async channel in request object"
+                                                                        (reset! *async-channel (or async-channel
```

Comment:
Why not make it a `promise`, then switch from `if-let` to `if (realized? *async-channel)`

Reply:
I think semantically, a _promise_ is a value that will eventually be supplied, whereas this use of an atom highlights that it _may_ be supplied.